core/platform/lf_nrf52_support.c

```diff
@@ -0,0 +1,310 @@
+    }
+    return 0;
+}
+
+/**
+ * @brief Do a busy-wait until a time instant
+ * 
+ * @param wakeup_time 
+ */
+
+static void lf_busy_wait_until(instant_t wakeup_time) {
+    instant_t now;
+    do {
+        lf_clock_gettime(&now);
+    } while (now < wakeup_time);
+}
+
+/**
+ * @brief Sleep until the given wakeup time. There are a couple of edge cases to consider
+ *  1. Wakeup time is already past
+ *  2. Implied sleep duration is below `LF_MAX_SLEEP_NS` threshold
+ *  3. Implied sleep duration is above `LF_MAX_SLEEP_NS` limit
+ * 
+ * @param wakeup_time The time instant at which to wake up.
+ * @return int 0 if sleep completed, or -1 if it was interrupted.
+ */
+int lf_sleep_until(instant_t wakeup_time) {
+    instant_t now;
+    lf_clock_gettime(&now);
+    interval_t duration = wakeup_time - now;
+    if (duration <= 0) {
+        return 0;
+    } else if (duration < LF_MIN_SLEEP_NS) {
+        lf_busy_wait_until(wakeup_time);
+        return 0;
+    } 
+
+    // The sleeping while loop continues until either:
+    // 1) A physical action is scheduled, resulting in a new event on the event queue
+    // 2) Sleep has completed successfully
+    bool sleep_next = true;
+    _lf_sleep_interrupted = false;
+    _lf_async_event = false;
+
+    do {
+        // Schedule a new timer interrupt unless we already have one pending
+        if (!_lf_sleep_interrupted) {
+            uint32_t curr_timer_val = nrfx_timer_capture(&g_lf_timer_inst, NRF_TIMER_CC_CHANNEL2);
+            uint32_t target_timer_val = 0;
+            // If the remaining sleep is longer than the limit, sleep for the maximum possible time.
+            if (duration > LF_MAX_SLEEP_NS) {
+                target_timer_val = curr_timer_val-1;
+                duration -= LF_MAX_SLEEP_NS;
+            } else {
+                target_timer_val = (uint32_t)(wakeup_time / 1000);
+                sleep_next = false;        
+            }
+            // init timer interrupt for sleep time
+            _lf_sleep_interrupted = true;
+            nrfx_timer_compare(&g_lf_timer_inst, NRF_TIMER_CC_CHANNEL2, target_timer_val, true);
+        }
+
+        // Leave critical section
+        lf_critical_section_exit();
+        
+        // wait for exception
+        __WFE();
+
+        // Enter critical section again
+        lf_critical_section_enter();
+
+        // Redo while loop and go back to sleep if:
+        //  1) We didnt have async event AND
+        //  2) We have more sleeps left OR the sleep didnt complete
+        // 
+        // This means we leave the sleep while if:
+        //  1) There was an async event OR
+        //  2) no more sleeps AND sleep not interrupted 
+    } while(!_lf_async_event && (sleep_next || _lf_sleep_interrupted));
+    
+    
+    
+    if (!_lf_async_event) {
+        return 0;
+    } else {
+        LF_PRINT_DEBUG("Sleep got interrupted...\n");
+        return -1;
+    }
+}
+
+int lf_critical_section_enter() {
```

Comment:
Comments needed for these functions.

Also, See my suggested implementation in the Arduino code for an alternative that will address this FIXME and also support nested critical sections.